eip/EIPS/eip-4337.md

```diff
@@ -343,19 +343,21 @@ Exceptions to the forbidden opcodes:
 2. `GAS` is allowed if followed immediately by one of { `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` }.
    (that is, making calls is allowed, using `gasleft()` or `gas` opcode directly is forbidden)
 
-### Reputation scoring and throttling/banning for paymasters
+### Reputation scoring and throttling/banning for paymasters, deployers and aggregators
 #### Rationale
 
-Paymasters might influence many UserOperations in the mempool.
-To prevent abuse, we throttle down (or completely ban for a period of time) a paymaster that causes invalidation of large number of UserOperations in the mempool.
-To prevent paymasters from "sybil-attack", we require a paymaster to stake with the system, and thus make such DoS attack very expensive.
+UserOperation's memory access rules prevent them from interfere with each other.
+But "global" entities - paymasters, deployers and aggregators are accessed by multiple UserOperations, and thus might influence them.
+To prevent abuse, we throttle down (or completely ban for a period of time) an entity that causes invalidation of large number of UserOperations in the mempool.
+To prevent such entities from "sybil-attack", we require them to stake with the system, and thus make such DoS attack very expensive.
 Note that this stake is never slashed, and can be withdrawn any time (after unstake delay)
+The only exemption from staking is if the entity doesn't use global storage
+(it is allowed to use storage specific to the account, such as mapping keyed by account address)
```

Comment:
"It is allowed to use storage specific to the account - any slots within the account, as well as any mapping keyed by the account address."

btw, this actually enables some useful unstaked paymasters.  E.g. a VerifyingPaymaster (which seems to be the most widely used use case so far) doesn't use storage at all, and even a TokenPaymaster could be written without accessing any global storage not keyed by the account address.  It's great if we can allow this safely.  Most paymasters won't need to stake anymore.

It does complicate the bundler a bit.  The bundler can no longer drop ops due to having an unstaked paymaster.  It must simulate the op, see that its paymaster validation stage doesn't access global storage, and only check for the stake if it does.

A paymaster could still invalidate many ops by depleting or withdrawing its deposit.  But luckily it's a cheap drop that doesn't require 2nd simulation.  These ops won't be included in the same bundle because the bundler already keeps track while building the bundle, and they'll be dropped the next time the bundler looks at them and sees that they require more gas than the paymaster can afford.

The paymaster could mess with the bundler's calculation by withdrawing its own deposit during validation (since `validatePaymasterUserOp` is no longer `view), but to do this it must access a storage slot in EntryPoint associated with itself (and not the account), so this paymaster will have to be staked and therefore would get itself banned.

Any other way an unstaked paymaster could invalidate other ops (other than storage access which we block for unstaked paymasters)?  If there aren't any, then we just improved paymasters significantly.

One thing we may need to clarify (since `validatePaymasterUserOp` is no longer `view`), is that `validatePaymasterUserOp` can read storage but not write to it.  We said the bundler will emulate the behavior of `staticcall` but I don't see where we wrote it in the EIP.  What we wrote is "The second (validatePaymasterUserOp), the paymaster may access storage slots of itself AND of the account, above." since we wrote it while it was still `view` and couldn't write.  Should we clarify that it's not allowed to write to storage?

Reply:
Another benefit of the above (if it's indeed safe) is that it'll remove some objections to high stake requirements.  If most paymasters don't need to stake, then maybe it's ok to require a higher stake from the rare ones who do.

Reply:
@drortirosh please see the questions in the above comment.  I think today's text changes still just use "storage access" and not make a distinction between reading and writing.  Should we clarify that entities are not allowed to write to storage outside their own, just read it, or are we ok with them writing as well?